### PR TITLE
[FLATCAP] Major jQuery and layout update

### LINKED ISSUE #681: Update jQuery version
Update jQuery dependency and stop using jQuery 1.x. This will coincide with communication to customers informing them of official de-support of IE8, IE9 and IE10, and reminding them that IE7 is not supported.

Plan of attack:
1. Pulsar branch to be created which updates jQuery dependency, initial impact to be documented as issue tickets and remediation work to happen ASAP
1. Experimental branches of CMS, XFP and CXM to be created which pull in the Pulsar update, initial findings regarding impact severity to be documented and discussed with Product Owners
1. Customers to be informed of the planned de-support of older IE versions, currently aiming for comms to go out at, or around the time of, the July Birmingham Academy
1. Product fixes to be documented, likely under an 'epic', and Product Owners to make a plan for remediation works to be completed by the end of 2019

## Migrate tool

https://blog.jquery.com/2016/05/19/jquery-migrate-1-4-1-released-and-the-path-to-jquery-3-0/

### LINKED ISSUE #1098: Disabled field values too light in Firefox
**Describe the bug**
Values with disabled fields in Firefox is adopting an incorrect text colour from the browser's forms.css default stylesheet.

**Expected behavior**
The text should use the standard base text colour and meet WCAG AA contrast when set against the disabled field background.

**Screenshots**
Screenshot 2019-12-02 at 13 42 17

**Desktop (please complete the following information):**
 - OS: MacOS
 - Browser: Firefox
 - Version: Observed in 70.0.1

### LINKED ISSUE #1138: Search icon is misaligned
Screenshot 2020-02-11 at 10 59 21

Issue exists in all browsers

### LINKED ISSUE #1140: Mobile focus order is incorrect
The focus order of the elements in the `header.toolbar` on mobile isn't ideal. As there isn't any relationship between the menu button, dropdowns and search form I'm not sure it's a strict failure (under 2.4.3 focus order or 1.3.2: Meaningful Sequence) but does warrant investigation.

Current focus order:

1. skip to content
2. menu button
3. search input
4. user dropdown
5. notifications

Ideal focus order:

1. skip to content
2. menu button
3. user dropdown
4. notifications
5. search input

### LINKED ISSUE #1185: Help side bar axe error
Fix AXE warning about duplicate landmarks. This is due to `tab__sidebar` and `tab-help-container` both having the same `aria-label`.

image

### PR DESCRIPTION
**Pull request by Stanton**
**This PR is closed to new commits, the `CMS/flatcap` branch is currently referenced by the `20.0.0` release of Jadu CMS and should not be deleted**

# Overview
* Update to jQuery 3.x. Closes #681 
* Increase contrast of text inside disabled fields. Closes #1098 
* Toolbar has been rebuilt with flexbox and appropriate source order. Closes #1140 
  * Fix position of search button. Closes #1138
* Update aria label for responsive help sidebar container. Closes #1185
* SelectWoo version updated
  * Fix height of select2 multiple selections
  * Fix height of multiple select2 instances used in the rules pattern
* Base template structure updated
* Remove positive tabindexes from elements which previously used them to manage tab order
* Panel markup updated (main structural elements)
* Cards can now use spans as well as divs for body
* Tweaks to actions bar and toolbar layouts following landmark layout changes
* Allow dropdown button items to be buttons as well as links
* Fix link contrast
  * in table rows
  * in sidebar
  * in naked lists
* CMS Module Page component fixes #1031
  * Fix CMS issue causing modules with no pages to be incorrectly unselected
  * Add indeterminate styling to restyled checkboxes

# Breaking change remediation
The following steps will need checking or implementing in Jadu products:

* Base layout
  * check position of all elements matches the base template outline
  * check all elements use the correct elements, aria attributes and classes
* package.json
  * Check all new dependencies are included
* index.js
  * check any deps previously controlled by bower are now pointing to node_modules
* main sass files
  * check any deps previously controlled by bower are now pointing to node_modules
  * Check shame files are using new 'scratch' name instead

# Base template outline
```<!DOCTYPE html>
<html lang="en" class="no-js">
    <head>
        <!-- head -->
    </head>
    <body>
        <a class="skip-link" href="#skip-target">Skip to main content</a>

        <div class="container">
            <div class="inner-wrapper">

                <!-- primary navigation -->

                <header class="toolbar">
                    <!-- mobile menu close button -->
                    <!-- search -->
                    <!-- site switcher (optional) -->
                    <!-- user menu -->
                    <!-- user guides (optional) -->
                    <!-- notifications (optional) -->
                    <!-- flash messages (optional) -->
                    <!-- breadcrumbs (optional) -->
                    
                    <h1 class="main-title">Main Title</h1>
                </header>

                <div class="content-main">
                    <div class="tabs__content">
                        <!-- ribbon (optional) -->
                        <nav class="nav-inline" aria-label="secondary">
                            <!-- tabs list (optional) -->
                        </nav>

                        <div class="tab__pane is-active">
                            <div class="tab__container">

                                <section class="actionsbar" aria-label="page actions">
                                    <div class="float--left">
                                        <!-- left actions (optional) -->
                                    </div>

                                    <div class="float--right">
                                        <!-- right actions (optional) -->
                                    </div>
                                </section>

                                <main class="tab__content" id="skip-target">
                                    <!-- tab content -->
                                </main>

                            </div>
                        </div>

                    </div>
                </div>

                <!-- empty help container which will be populated at responsive modes -->
                <aside class="tab-help-container" aria-label="additional help">
                    <div class="tab-help">
                    </div>
                </aside>

                <!-- empty modal which can be populated by javascript -->
                <div class="modal" role="dialog" aria-modal="true" aria-hidden="true">
                </div>

            </div>
        </div>

        <footer class="footer">
            <!-- footer links -->
        </footer>

    </body>
</html>
```

### CODE REVIEW

File: js/PulsarFormComponent.js [hunk 1]
```diff
@@ -1,10 +1,11 @@
 'use strict';
 
 var $ = require('jquery'),
-    TimePickerComponent = require('./TimePickerComponent');
+TimePickerComponent = require('./TimePickerComponent');
```
Comment (jamesjacobs) [inline js/PulsarFormComponent.js]: ~tabbing~ Actually it looks like this branch also contains the JQ3 changes, might be better to merge that branch first so the JQ3 changes are separated out as this PR is big enough already!

File: stylesheets/_component.toolbar.scss [hunk 2]
```diff
@@ -26,15 +54,14 @@
         left: $padding-small-horizontal;
         position: absolute;
         right: $padding-small-horizontal;
-        top: 60px;
+        // top: 60px;
```
Comment (jamesjacobs) [inline stylesheets/_component.toolbar.scss]: Remove?

File: stylesheets/_component.toolbar.scss [hunk 3]
```diff
@@ -46,17 +73,18 @@
 }
 
 .nav-user {
-    float: right;
+    // float: left;
```
Comment (jamesjacobs) [inline stylesheets/_component.toolbar.scss]: remove?

File: stylesheets/_component.toolbar.scss [hunk 4]
```diff
@@ -46,17 +73,18 @@
...
 
     .btn {
         font-size: $font-size-base;
     }
 
     .dropdown {
-        float: left;
-        margin-left: 20px;
-        min-width: 60px;
+        // float: left;
```
Comment (jamesjacobs) [inline stylesheets/_component.toolbar.scss]: remove these commented lines?

File: stylesheets/_component.toolbar.scss [hunk 5]
```diff
@@ -90,7 +118,7 @@
 
 .toolbar-icon,
 %toolbar-icon {
-    float: right;
+    // float: left;
```
Comment (jamesjacobs) [inline stylesheets/_component.toolbar.scss]: remove?

File: stylesheets/_deprecated.styles.scss [hunk 6]
```diff
@@ -468,42 +468,42 @@ $rich-list-item-text-selected-color:  color(white) !default;
...
-.grid-col {
-    margin-top: gutter();
-
-    &:last-child {
-        @include respond-min($screen-phone) {
-            @include last;
-        }
-    }
-}
+// $susy: (
```
Comment (jamesjacobs) [inline stylesheets/_deprecated.styles.scss]: Remove?

File: js/NavMainComponent.js [hunk 7]
```diff
@@ -53,8 +53,8 @@ NavMainComponent.prototype.init = function () {
         if ($self.text() === 'Menu') {
             $self.text('Close');
             $self.attr('aria-expanded', 'true');
-            component.$brandingLink.attr('tabindex', '3');
-            component.$primaryNavLinks.attr('tabindex', '3');
+            component.$brandingLink.attr('tabindex', '0');
```
Comment (jamesjacobs) [inline js/NavMainComponent.js]: Why do we need to set a zero tabindex here and below?

File: stylesheets/_layout.settings.scss [hunk 8]
```diff
@@ -227,11 +227,11 @@
  * Form toggle switches
  */
 
-.tab__settings .form__group--toggle > .control__label {
-    width: 100%;
+.tab__settings .form__group--toggle span.control__label {
+    width: calc(100% - 4em);
```
Comment (jamesjacobs) [inline stylesheets/_layout.settings.scss]: 4em?

File: stylesheets/_layout.settings.scss [hunk 9]
```diff
@@ -227,11 +227,11 @@
...
 
     @include respond-min($screen-small) {
-        padding-right: calc(4em + #{$padding-small-horizontal * 2});
+        // padding-right: calc(4em + #{$padding-small-horizontal * 2});
```
Comment (jamesjacobs) [inline stylesheets/_layout.settings.scss]: remove?

File: stylesheets/_layout.settings.scss (hunk 8 above)
Comment (Stanton) [reply stylesheets/_layout.settings.scss]: = width of `.controls` for toggles

File: js/NavMainComponent.js (hunk 7 above)
Comment (Stanton) [reply js/NavMainComponent.js]: We switch between tabindex `0` and `-1`, we could remove the tabindex now we're not switching between values greater than 0, however toggling between these `0` and `-1` was a bit simpler to keep track of in code and through tests.

File: stylesheets/_component.breadcrumb.scss [hunk 10]
```diff
@@ -1,6 +1,8 @@
 .breadcrumb {
+    clear: both;
     color: color(gray, dark); // must be AA 4.5:1
-    margin-bottom: 11px;
+    margin: 0 0 ($line-height-base / 2);
+    width: 100%;
```
Comment (jamesjacobs) [inline stylesheets/_component.breadcrumb.scss]: Needs checking to see if still needed following flash changes (which included other `header.toolbar` spacing tweaks.

File: stylesheets/_component.flash.scss [hunk 11]
```diff
@@ -111,8 +111,13 @@
 }
 
 .flash-container {
-    left: $nav-primary-width;
+    left: 0;
```
Comment (jamesjacobs) [inline stylesheets/_component.flash.scss]: `flash-container` is no longer absolutely positioned so this change should not be needed

File: stylesheets/_component.flash.scss [hunk 12]
```diff
@@ -111,8 +111,13 @@
...
     position: fixed;
     right: 0;
     z-index: $zindex-flash;
+
+    // accommodate sidebar navigation
+    @include respond-min($screen-desktop) {
+        left: $nav-primary-width;
```
Comment (jamesjacobs) [inline stylesheets/_component.flash.scss]: `flash-container` is no longer absolutely positioned so this change should not be needed

File: stylesheets/_component.lists.scss [hunk 13]
```diff
@@ -13,6 +13,10 @@
         color: color(primary, alt);
         text-decoration: none;
     }
+
+    & > a:focus {
+        color: color(text);
```
Comment (jamesjacobs) [inline stylesheets/_component.lists.scss]: Is this needed in addition to standard focus styles?

File: stylesheets/_component.navigation.scss [hunk 14]
```diff
@@ -179,8 +179,9 @@
     border-radius: 4px;
     color: $nav-color-dark;
     display: inline-block;
+    height: 34px;
```
Comment (jamesjacobs) [inline stylesheets/_component.navigation.scss]: Is there a need to set height on a button here?

File: stylesheets/_component.search.scss [hunk 15]
```diff
@@ -1,19 +1,8 @@
 .main-search {
     border: 1px solid $input-border;
     border-radius: 4px;
-    clear: both;
-    display: block;
-    margin: 20px 0;
+    height: 34px;
```
Comment (jamesjacobs) [inline stylesheets/_component.search.scss]: Unsure why these styles have been removed.

File: views/pulsar/layouts/base.html.twig [hunk 16]
```diff
@@ -58,76 +58,79 @@
     <a class="skip-link" href="#skip-target" tabindex="1">Skip to main content</a>
 
 {% block body %}
-    <div class="container" id="top">
+    <div role="main" class="container" id="top">
```
Comment (jamesjacobs) [inline views/pulsar/layouts/base.html.twig]: This is no longer correct, `tab__content` is now `role="main"`

File: views/pulsar/layouts/base.html.twig [hunk 17]
```diff
@@ -58,76 +58,79 @@
...
-                                "href": "#"
-                            })
-                        ]
-                    })
-                }}
-                {% endblock %}
-                </div>
-
-                {% include '@pulsar/pulsar/components/nav_help.html.twig' %}
+            {% include '@pulsar/pulsar/components/flash.html.twig' %}
```
Comment (jamesjacobs) [inline views/pulsar/layouts/base.html.twig]: All these changes in base.html.twig look to be out of sync with the more recent changes in develop.

File: views/pulsar/layouts/base.html.twig (hunk 16 above)
Comment (Stanton) [reply views/pulsar/layouts/base.html.twig]: Clarify this for me, as I have

```
<main class="tab__content" id="skip-target">
```

File: stylesheets/_component.navigation.scss (hunk 14 above)
Comment (Stanton) [reply stylesheets/_component.navigation.scss]: I was struggling to set the height of toolbar elements to a consistent height

File: stylesheets/_component.lists.scss (hunk 13 above)
Comment (Stanton) [reply stylesheets/_component.lists.scss]: May not be required now based on #1186

**Pull request merged**